src/core/tsi/ssl_transport_security.cc

```diff
@@ -1450,9 +1472,10 @@ static tsi_result ssl_handshaker_next(
   if (ssl_handshaker_get_result(impl) == TSI_HANDSHAKE_IN_PROGRESS) {
     *handshaker_result = nullptr;
   } else {
-    size_t unused_bytes_size = received_bytes_size - bytes_consumed;
-    const unsigned char* unused_bytes =
-        unused_bytes_size == 0 ? nullptr : received_bytes + bytes_consumed;
+    unsigned char* unused_bytes = nullptr;
```

Comment:
I did not understand this part. So with ssl_bytes_remaining, another read from SSL is done. However, is it possible that received_bytes_size is larger than bytes_consumed? The old code seems to assume so and in the new code that part of data is just ignored?

Reply:
The old code had the following bug in the implementation of `ssl_handshaker_process_bytes_from_peer`. It computed `bytes_consumed` as the number of bytes written to (the write BIO of) `impl->bio`. Those bytes are then moved to (the read BIO of) `impl->ssl` when `SSL_do_handshake` is called, so `bytes_consumed` will *always* be equal to `received_bytes_size`. This bug was never caught because the TLS 1.2 RFC explicitly prohibits any application data from being appended to a handshake message (i.e. there are never any unused bytes for TLS 1.2).

The new code computes `unused_bytes_size` by reading from the correct BIO, namely (the read buffer of) `impl->ssl`. Does that make sense?

Additionally (and this might be what you were asking), there *should* be a check that `unused_bytes_size <= received_bytes_size` after the call to `ssl_bytes_remaining`. I'll add this in a separate commit.